File: bindings/fortran/test/conformance.f90
! Uniform facade conformance suite for the Fortran binding.
!
! Instantiates the family C1-C10 contract for Fortran against a live
! libdictenstein shared library:
!   C1 identity/version      test_identity
!   C2 lifecycle/ownership   test_lifecycle       (idempotent close)
!   C3 error mapping + text  test_error_message   (INVALID_UTF8 + thread-local)
!   C5 CRUD + substring      test_crud, test_substring
!   C6 text domains          test_unicode         (U+00E9 round-trip)
!   C7 batch edges           test_batch_sizes     (0/1/255/256/257)
!   persistence round-trip   test_persistence     (create/checkpoint/reopen)
!
! Run from bindings/fortran with the shared library on the linker and loader
! search paths, e.g.:
!   LIBRARY_PATH=../../target/debug LD_LIBRARY_PATH=../../target/debug \
!     fpm test --link-flag "-L../../target/debug"
program test_conformance
  use, intrinsic :: iso_c_binding
  use vinary_tree_libdictenstein
  implicit none
  integer :: folded_entries = 0

  call test_identity()
  call test_lifecycle()
  call test_error_message()
  call test_crud()
  call test_unicode()
  call test_batch_sizes()
  call test_substring()
  call test_persistence()
  call test_entries()
  print *, "fortran conformance: all checks passed"

contains

  subroutine check(condition, label)
    logical, intent(in) :: condition
    character(len=*), intent(in) :: label
    if (.not. condition) then
      print *, "FAIL: ", label
      error stop 1
    end if
  end subroutine

  ! C1: identity constants are surfaced through the facade.
  subroutine test_identity()
    call check(abi_version() == 1, "abi_version == 1")
    call check(api_revision() == 6, "api_revision == 6")
  end subroutine

  ! C2: construction reports its kind; close is idempotent (no double free).
  subroutine test_lifecycle()
    type(dictionary) :: d
    integer(c_int32_t) :: st
    call new_dynamic_dawg(d, status=st)
    call check(st == ldict_ok, "dynamic dawg constructed")
    call check(d%kind() == dynamic_dawg_kind, "kind == dynamic dawg")
    call check(iand(d%capabilities(), can_insert) /= 0, "advertises insert")
    call d%close()
    call d%close()
    call check(.true., "double close is a no-op")
  end subroutine

  ! C3: a failing call yields a non-OK status and a non-empty thread-local
  ! message. A lone 0xFF byte is invalid UTF-8 for a Unicode-scalar backend.
  subroutine test_error_message()
    type(dictionary) :: d
    integer(c_int32_t) :: st
    logical :: ins
    character(kind=c_char, len=:), allocatable :: msg
    call new_dynamic_dawg(d, status=st)
    call d%put_text(achar(255), inserted=ins, status=st)
    call check(st /= ldict_ok, "invalid utf-8 insert rejected")
    msg = last_error_message()
    call check(len(msg) > 0, "last_error_message is non-empty after failure")
    call d%close()
  end subroutine

  ! C5: exact CRUD and the optional-u64 value channel (present and absent).
  subroutine test_crud()
    type(dictionary) :: d
    integer(c_int32_t) :: st
    logical :: ins, rem, fnd, hv
    integer(c_int64_t) :: val
    call new_dynamic_dawg(d, status=st)
    call d%put_text("cat", value=1_c_int64_t, inserted=ins, status=st)
    call check(ins, "insert cat")
    call d%put_text("cot", value=2_c_int64_t, inserted=ins)
    call check(ins, "insert cot")
    call d%put_text("cut", inserted=ins)
    call check(ins, "insert cut valueless")
    call check(int(d%length()) == 3, "length == 3")
    call check(d%contains_text("cat"), "contains cat")
    call check(.not. d%contains_text("dog"), "not contains dog")
    call d%get_text("cot", found=fnd, value=val, has_value=hv, status=st)
    call check(fnd .and. hv .and. val == 2, "get cot == 2")
    call d%get_text("cut", found=fnd, value=val, has_value=hv)
    call check(fnd .and. (.not. hv), "cut found valueless")
    call d%remove_text("cot", removed=rem, status=st)
    call check(rem, "remove cot")
    call check(.not. d%contains_text("cot"), "cot gone")
    call check(int(d%length()) == 2, "length == 2 after remove")
    call d%close()
  end subroutine

  ! C6: a multibyte Unicode scalar (U+00E9, UTF-8 C3 A9) round-trips.
  subroutine test_unicode()
    type(dictionary) :: d
    integer(c_int32_t) :: st
    logical :: ins
    character(kind=c_char, len=2) :: eacute
    eacute = achar(195)//achar(169)
    call new_dynamic_dawg(d, status=st)
    call d%put_text(eacute, inserted=ins, status=st)
    call check(st == ldict_ok, "insert e-acute ok")
    call check(ins, "e-acute inserted")
    call check(d%contains_text(eacute), "contains e-acute")
    call d%close()
  end subroutine

  ! C7: single-crossing batch insert across the paging edge cases.
  subroutine test_batch_sizes()
    integer :: sizes(5), s, i, k
    integer(c_int32_t) :: st
    integer(c_size_t) :: n
    type(dictionary) :: d
    character(len=16), allocatable :: terms(:)
    sizes = [0, 1, 255, 256, 257]
    do s = 1, size(sizes)
      k = sizes(s)
      call new_dynamic_dawg(d, status=st)
      allocate(terms(k))
      do i = 1, k
        write(terms(i), '(A,I0)') "t", i
      end do
      call d%put_all(terms, inserted=n, status=st)
      call check(st == ldict_ok, "batch status ok")
      call check(int(n) == k, "batch inserted == size")
      call check(int(d%length()) == k, "length == batch size")
      deallocate(terms)
      call d%close()
    end do
  end subroutine

  ! C5 substring: SCDAWG advertises SUBSTRING and answers membership/frequency.
  subroutine test_substring()
    type(dictionary) :: d
    integer(c_int32_t) :: st
    logical :: ins
    integer(c_size_t) :: freq
    call new_scdawg(d, status=st)
    call check(st == ldict_ok, "scdawg constructed")
    call check(iand(d%capabilities(), can_substring) /= 0, "scdawg advertises substring")
    call d%put_text("cat", inserted=ins, status=st)
    call d%put_text("cot", inserted=ins)
    call d%put_text("cut", inserted=ins)
    call check(d%contains_substring("at"), "substring at present")
    call check(.not. d%contains_substring("xyz"), "substring xyz absent")
    freq = d%substring_frequency("t")
    call check(int(freq) == 3, "frequency t == 3")
    call d%close()
  end subroutine

  ! Persistence: create/checkpoint/free/reopen preserves committed state.
  subroutine test_persistence()
    type(dictionary) :: d
    integer(c_int32_t) :: st
    logical :: ins, fnd, hv
    integer(c_int64_t) :: val
    integer :: clock
    character(len=80) :: path
    call system_clock(clock)
    write(path, '(A,I0,A)') "/tmp/vt_fortran_conformance_", clock, ".part"
    call create_persistent_artrie(d, trim(path), status=st)
    call check(st == ldict_ok, "persistent create ok")
    call check(d%kind() == persistent_artrie_kind, "kind == persistent artrie")
    call check(iand(d%capabilities(), can_checkpoint) /= 0, "advertises checkpoint")
    call d%put_text("cat", value=1_c_int64_t, inserted=ins, status=st)
    call d%put_text("cot", value=2_c_int64_t, inserted=ins)
    call d%checkpoint(status=st)
    call check(st == ldict_ok, "checkpoint ok")
    call d%close()
    call open_persistent_artrie(d, trim(path), status=st)
    call check(st == ldict_ok, "persistent reopen ok")
    call d%get_text("cot", found=fnd, value=val, has_value=hv, status=st)
    call check(fnd .and. hv .and. val == 2, "reopened cot == 2")
    call d%close()
  end subroutine

  ! entries-v1: one immutable ordered revision, deep host-owned batches,
  ! deterministic early close, and exact unit/value preservation.
  subroutine test_entries()
    type(dictionary) :: d
    type(dictionary_entry_cursor) :: cursor
    type(dictionary_entries_metadata) :: metadata
    type(dictionary_entry_batch) :: batch
    type(dictionary_entry) :: saved
    integer(c_int32_t) :: st
    integer(c_int64_t) :: high_bit
    integer(c_int64_t) :: one(1)
    logical :: ins
    character(kind=c_char, len=5) :: cafe
    character(kind=c_char, len=2) :: raw

    cafe = "caf"//achar(195)//achar(169)
    call new_dynamic_dawg(d, domain=vt_unit_unicode_scalar, status=st)
    call d%put_text("cat", inserted=ins)
    call d%put_text(cafe, value=-1_c_int64_t, inserted=ins)
    call d%put_text("", value=0_c_int64_t, inserted=ins)
    call d%open_entries(cursor, status=st, max_entries=1_c_size_t)
    call check(st == ldict_ok, "entries cursor opened")
    metadata = cursor%metadata()
    call check(metadata%unit_domain == vt_unit_unicode_scalar, "entries unicode domain")
    call check(metadata%value_domain == 1, "entries optional-u64 values")
    call check(metadata%has_exact_length .and. metadata%exact_length == 3, "entries exact length")
    call check(metadata%has_snapshot_identity, "entries snapshot identity")
    call d%put_text("dog", value=4_c_int64_t, inserted=ins)

    call cursor%next_batch(batch, max_entries=1_c_size_t, status=st)
    call check(st == ldict_ok .and. batch%count == 1, "entries first batch")
    call check(allocated(batch%entries(1)%scalars), "entries scalar payload allocated")
    call check(size(batch%entries(1)%scalars) == 0, "entries empty key first")
    call check(batch%entries(1)%has_value .and. batch%entries(1)%value == 0, "entries zero value")
    saved = batch%entries(1)

    call cursor%next_batch(batch, max_entries=1_c_size_t, status=st)
    call check(batch%count == 1 .and. size(batch%entries(1)%scalars) == 4, "entries cafe scalar length")
    call check(all(batch%entries(1)%scalars == [99, 97, 102, 233]), "entries cafe scalar values")
    call check(batch%entries(1)%has_value .and. batch%entries(1)%value == -1, "entries full-width value bits")
    call check(allocated(saved%scalars) .and. size(saved%scalars) == 0, "entries batch copy survives refill")

    call cursor%next_batch(batch, max_entries=1_c_size_t, status=st)
    call check(batch%count == 1 .and. all(batch%entries(1)%scalars == [99, 97, 116]), "entries cat third")
    call check(.not. batch%entries(1)%has_value, "entries present-unvalued preserved")
    call cursor%next_batch(batch, max_entries=1_c_size_t, status=st)
    call check(st == ldict_ok .and. batch%count == 0, "entries snapshot excludes later mutation")
    call cursor%close(st)
    call check(st == ldict_ok, "entries cursor close is idempotent after end")

    folded_entries = 0
    call d%fold_entries(stop_after_first_entry, status=st, max_entries=1_c_size_t)
    call check(st == ldict_ok .and. folded_entries == 1, "entries fold early cancellation")
    call check(d%contains_text("cat"), "dictionary usable after early entries close")
    call d%close()

    raw = achar(0)//achar(255)
    call new_dynamic_dawg(d, domain=vt_unit_byte, status=st)
    call d%put_text("", inserted=ins)
    call d%put_text(raw, value=-1_c_int64_t, inserted=ins)
    call d%open_entries(cursor, status=st)
    call cursor%next_batch(batch, status=st)
    call check(batch%count == 2, "entries byte count")
    call check(size(batch%entries(1)%bytes) == 0, "entries empty byte key first")
    call check(size(batch%entries(2)%bytes) == 2, "entries arbitrary byte length")
    call check(iachar(batch%entries(2)%bytes(1)) == 0 .and. &
      iachar(batch%entries(2)%bytes(2)) == 255, "entries arbitrary byte values")
    call check(batch%entries(2)%value == -1, "entries byte full-width value bits")
    call cursor%close()
    call d%close()

    high_bit = not(huge(0_c_int64_t))
    call new_dynamic_dawg(d, domain=vt_unit_u64, status=st)
    one = [-1_c_int64_t]
    call d%put_u64(one, value=-1_c_int64_t, inserted=ins)
    one = [0_c_int64_t]
    call d%put_u64(one, inserted=ins)
    one = [high_bit]
    call d%put_u64(one, value=0_c_int64_t, inserted=ins)
    call d%open_entries(cursor, status=st)
    call cursor%next_batch(batch, status=st)
    call check(batch%count == 3, "entries u64 count")
    call check(batch%entries(1)%tokens(1) == 0, "entries u64 zero first")
    call check(batch%entries(2)%tokens(1) == high_bit, "entries u64 high bit second")
    call check(batch%entries(3)%tokens(1) == -1, "entries u64 max third")
    call check(batch%entries(3)%has_value .and. batch%entries(3)%value == -1, &
      "entries u64 max value bits")
    call cursor%close()
    call d%close()
  end subroutine

  subroutine stop_after_first_entry(batch, stop)
    type(dictionary_entry_batch), intent(in) :: batch
    logical, intent(out) :: stop
    folded_entries = folded_entries + int(batch%count)
    stop = .true.
  end subroutine

end program test_conformance

File: bindings/ocaml/test/conformance.ml
(* Uniform facade conformance suite for the OCaml binding.

   Instantiates the family C1-C10 contract for OCaml against a live
   libdictenstein shared library. It needs only libdictenstein and the canonical
   fixture, never a liblevenshtein transducer, so it pins the *producer* ABI in
   isolation.

     C1  identity + kind/capabilities per backend
     C2  idempotent close + free-order independence
     C3  failure raised (+ non-empty message) for INVALID_UTF8 / DOMAIN_MISMATCH
         / IO_ERROR (the facade raises [Failure msg]; the numeric status is not
         carried, so status-code granularity is N/A)
     C4  canonical fixture replay (all four backends)
     C5  CRUD + value + batch + substring; capability-derived assertions
     C6  precomposed/combining/multibyte, byte-domain NUL + invalid UTF-8, u64 0/MAX
     C7  batch sizes 0/1/255/256/257/1000 (put_many)
     C8  CRUD op-script vs a Hashtbl oracle; substring vs a naive oracle
     C9  leak discipline (>=10k cycles, RSS bounded)
     C10 concurrency: independent per-domain dictionaries + readers during a
         writer (OCaml 5 Domain)

   Run (with the cdylib on the loader/linker path):
     LIBRARY_PATH=target/release LD_LIBRARY_PATH=target/release \
       dune exec bindings/ocaml/test/conformance.exe -- bindings/canonical_fixture.json *)

module D = Vinary_tree_libdictenstein
module I = Vinary_tree_interop
module U = Yojson.Safe.Util

let failures = ref 0
let check condition message =
  if not condition then begin
    Printf.eprintf "FAIL: %s\n" message;
    incr failures
  end

(* Capability bits (the LDICT_CAP_ family). *)
let cap_read = 1L
let cap_insert = 2L
let cap_remove = 4L
let cap_clear = 8L
let cap_compact = 16L
let cap_substring = 32L
let cap_checkpoint = 64L
let has caps bit = Int64.logand caps bit <> 0L

(* --------------------------------------------------------------------- *)
(* fixture (C4)                                                          *)
(* --------------------------------------------------------------------- *)

let optval json = match json with `Null -> None | v -> Some (Int64.of_int (U.to_int v))

let fixture_path =
  if Array.length Sys.argv > 1 then Sys.argv.(1)
  else
    let candidates =
      [ "bindings/canonical_fixture.json"; "../canonical_fixture.json";
        "../../canonical_fixture.json"; "../../../bindings/canonical_fixture.json" ]
    in
    (try List.find Sys.file_exists candidates
     with Not_found -> "bindings/canonical_fixture.json")

let root = Yojson.Safe.from_file fixture_path
let entries =
  U.member "entries" root |> U.to_list
  |> List.map (fun e -> (U.to_string (U.member "term" e), optval (U.member "value" e)))
  |> Array.of_list
let size = U.to_int (U.member "size" root)

let assert_fixture_reads dictionary =
  check (D.length dictionary = size) "fixture size";
  List.iter
    (fun case ->
      let term = U.to_string (U.member "term" case) in
      check (D.contains dictionary term = U.to_bool (U.member "expected" case)) ("contains " ^ term))
    (U.to_list (U.member "contains" root));
  List.iter
    (fun case ->
      let term = U.to_string (U.member "term" case) in
      let { D.found; value } = D.get dictionary term in
      check (found = U.to_bool (U.member "found" case)) ("get.found " ^ term);
      check (value = optval (U.member "value" case)) ("get.value " ^ term))
    (U.to_list (U.member "get" root))

(* --------------------------------------------------------------------- *)
(* C1 identity/version                                                   *)
(* --------------------------------------------------------------------- *)

let c1 () =
  check (D.abi_version () = 1) "abi version == 1";
  check (D.api_revision () = 4) "api revision == 4";
  let dawg = D.dynamic_dawg () in
  check (D.kind dawg = 1) "dawg kind";
  let caps = D.capabilities dawg in
  check (has caps cap_insert && has caps cap_remove && has caps cap_clear && has caps cap_compact) "dawg caps";
  check (not (has caps cap_substring) && not (has caps cap_checkpoint)) "dawg lacks substring/checkpoint";
  D.close dawg;
  let dat = D.double_array_trie [| ("x", None) |] in
  check (D.kind dat = 2) "dat kind";
  check (has (D.capabilities dat) cap_read) "dat read";
  D.close dat;
  let scdawg = D.scdawg () in
  check (D.kind scdawg = 3) "scdawg kind";
  check (has (D.capabilities scdawg) cap_substring) "scdawg substring";
  D.close scdawg

(* --------------------------------------------------------------------- *)
(* C2 lifecycle/ownership                                                *)
(* --------------------------------------------------------------------- *)

let c2 () =
  let dawg = D.dynamic_dawg () in
  ignore (D.put dawg "a" None);
  D.close dawg;
  D.close dawg; (* idempotent *)
  let dawgs = Array.init 4 (fun i -> let d = D.dynamic_dawg () in
                             ignore (D.put d (Printf.sprintf "term%d" i) (Some (Int64.of_int i))); d) in
  List.iter (fun i -> D.close dawgs.(i)) [ 2; 0; 3; 1 ]

(* --------------------------------------------------------------------- *)
(* C3 error mapping (facade raises Failure with a message)              *)
(* --------------------------------------------------------------------- *)

let raises_with_message thunk =
  try thunk (); (false, "") with
  | Failure message -> (true, message)
  | _ -> (true, "?")

let c3 () =
  let dawg = D.dynamic_dawg ~domain:I.Unicode_scalar () in
  let raised, message = raises_with_message (fun () -> ignore (D.put dawg "\xff" None)) in
  check (raised && String.length message > 0) "invalid utf8 raises with message";
  let raised, _ = raises_with_message (fun () -> ignore (D.put_u64 dawg [| 1L; 2L |] None)) in
  check raised "domain mismatch raises";
  D.close dawg;
  let path = Filename.temp_file "ldict-ocaml" ".part" in
  Sys.remove path;
  let raised, message = raises_with_message (fun () -> ignore (D.open_persistent_artrie path)) in
  check (raised && String.length message > 0) "io error raises with message"

(* --------------------------------------------------------------------- *)
(* C4 canonical fixture replay                                           *)
(* --------------------------------------------------------------------- *)

let c4 () =
  let dawg = D.dynamic_dawg () in
  check (D.put_many dawg entries = size) "dawg batch count";
  assert_fixture_reads dawg;
  D.close dawg;
  let dat = D.double_array_trie entries in
  assert_fixture_reads dat;
  D.close dat;
  let path = Filename.temp_file "ldict-ocaml-c4" ".part" in
  Sys.remove path;
  let art = D.create_persistent_artrie path in
  Array.iter (fun (term, value) -> ignore (D.put art term value)) entries;
  assert_fixture_reads art;
  D.close art;
  List.iter (fun suffix -> try Sys.remove (path ^ suffix) with _ -> ()) [ ""; ".wal"; ".wlock" ];
  let scdawg = D.scdawg () in
  Array.iter (fun (term, value) -> ignore (D.put scdawg term value)) entries;
  List.iter
    (fun case ->
      let pattern = U.to_string (U.member "pattern" case) in
      check (D.substring_frequency scdawg pattern = U.to_int (U.member "expected" case)) ("frequency " ^ pattern))
    (U.to_list (U.member "substring_frequency" root));
  List.iter
    (fun case ->
      let pattern = U.to_string (U.member "pattern" case) in
      check (D.contains_substring scdawg pattern = U.to_bool (U.member "expected" case)) ("contains_substring " ^ pattern))
    (U.to_list (U.member "substring_contains" root));
  D.close scdawg

(* --------------------------------------------------------------------- *)
(* C5 CRUD + value + batch + substring; capability-derived assertions    *)
(* --------------------------------------------------------------------- *)

let c5 () =
  let dawg = D.dynamic_dawg () in
  check (D.put dawg "cat" (Some 1L)) "insert cat";
  check (not (D.put dawg "cat" (Some 1L))) "idempotent insert";
  check ((D.get dawg "cat").value = Some 1L) "get cat";
  check (D.remove dawg "cat") "remove cat";
  check (not (D.remove dawg "cat")) "second remove";
  check (not (D.contains dawg "cat")) "cat gone";
  for i = 0 to 49 do ignore (D.put dawg (Printf.sprintf "t%d" i) (Some (Int64.of_int i))) done;
  for i = 0 to 49 do if i mod 2 = 0 then check (D.remove dawg (Printf.sprintf "t%d" i)) "remove even" done;
  ignore (D.compact dawg);
  check (D.length dawg = 25) "compact size";
  check ((D.get dawg "t1").value = Some 1L) "t1 survives";
  check (not (D.contains dawg "t0")) "t0 gone";
  D.close dawg;
  let scdawg = D.scdawg () in
  ignore (D.put scdawg "cat" (Some 1L));
  ignore (D.put scdawg "cot" (Some 2L));
  check (D.substring_frequency scdawg "t" = 2) "freq t == 2";
  check (D.put scdawg "cut" None) "insert cut";
  check (D.substring_frequency scdawg "t" = 3) "freq t == 3";
  D.close scdawg;
  let dat = D.double_array_trie [| ("x", None) |] in
  let caps = D.capabilities dat in
  check (not (has caps cap_insert) && not (has caps cap_remove)
         && not (has caps cap_clear) && not (has caps cap_compact)) "dat capability-derived reject";
  D.close dat

(* --------------------------------------------------------------------- *)
(* C6 text domains and values                                            *)
(* --------------------------------------------------------------------- *)

let c6 () =
  let dawg = D.dynamic_dawg () in
  check (D.put dawg "caf\xc3\xa9" (Some 7L)) "precomposed insert";     (* café, precomposed U+00E9 *)
  check (D.put dawg "\xf0\x9f\xa6\x80" (Some 255L)) "emoji insert";    (* crab, 4-byte scalar *)
  check (D.contains dawg "caf\xc3\xa9") "precomposed contains";
  check ((D.get dawg "\xf0\x9f\xa6\x80").value = Some 255L) "emoji value";
  D.close dawg;
  let dawg = D.dynamic_dawg () in
  let precomposed = "caf\xc3\xa9" and combining = "cafe\xcc\x81" in
  check (D.put dawg precomposed (Some 1L)) "precomposed distinct";
  check (D.put dawg combining (Some 2L)) "combining distinct";
  check (D.length dawg = 2) "distinct scalar sequences";
  check ((D.get dawg precomposed).value = Some 1L) "precomposed value";
  check ((D.get dawg combining).value = Some 2L) "combining value";
  D.close dawg;
  let dawg = D.dynamic_dawg ~domain:I.Byte () in
  check (D.put dawg "a\x00b" (Some 1L)) "embedded NUL insert";
  check (D.put dawg "\xff\xfe" (Some 2L)) "invalid utf8 byte insert";
  check (D.contains dawg "a\x00b") "embedded NUL contains";
  check ((D.get dawg "\xff\xfe").value = Some 2L) "invalid utf8 byte value";
  D.close dawg;
  let dawg = D.dynamic_dawg ~domain:I.U64 () in
  check (D.put_u64 dawg [| 1L; 2L; 3L |] (Some 0L)) "u64 value 0 insert";
  check (D.put_u64 dawg [| 9L |] (Some Int64.minus_one)) "u64 value MAX insert"; (* -1L = 2^64-1 in u64 bits *)
  check ((D.get_u64 dawg [| 1L; 2L; 3L |]).value = Some 0L) "u64 value 0";
  check ((D.get_u64 dawg [| 9L |]).value = Some Int64.minus_one) "u64 value MAX";
  D.close dawg

(* --------------------------------------------------------------------- *)
(* C7 batch / paging edges                                               *)
(* --------------------------------------------------------------------- *)

let c7 () =
  List.iter
    (fun batch_size ->
      let dawg = D.dynamic_dawg () in
      let batch = Array.init batch_size (fun i -> (Printf.sprintf "t%d" i, Some (Int64.of_int i))) in
      check (D.put_many dawg batch = batch_size) (Printf.sprintf "batch %d count" batch_size);
      check (D.length dawg = batch_size) (Printf.sprintf "batch %d size" batch_size);
      if batch_size > 0 then begin
        check ((D.get dawg "t0").value = Some 0L) (Printf.sprintf "batch %d first" batch_size);
        check ((D.get dawg (Printf.sprintf "t%d" (batch_size - 1))).value = Some (Int64.of_int (batch_size - 1)))
          (Printf.sprintf "batch %d last" batch_size)
      end;
      D.close dawg)
    [ 0; 1; 255; 256; 257; 1000 ]

(* --------------------------------------------------------------------- *)
(* C8 property-based testing vs an in-language oracle (deterministic LCG) *)
(* --------------------------------------------------------------------- *)

let make_rng seed =
  let state = ref seed in
  fun n ->
    state := Int64.add (Int64.mul !state 6364136223846793005L) 1442695040888963407L;
    let hi = Int64.shift_right_logical !state 33 in
    Int64.to_int (Int64.rem hi (Int64.of_int n))

let c8_crud () =
  let rng = make_rng 0xC0FFEEL in
  let keys = Array.init 40 (fun i -> Printf.sprintf "k%d" i) in
  let oracle : (string, int64 option) Hashtbl.t = Hashtbl.create 64 in
  let dawg = D.dynamic_dawg () in
  for _ = 1 to 3000 do
    let key = keys.(rng 40) in
    let present = Hashtbl.mem oracle key in
    let op = rng 100 in
    if op < 50 then begin
      let value = if rng 2 = 0 then None else Some (Int64.of_int (rng 1000000000)) in
      check (D.put dawg key value = not present) "crud insert changed";
      Hashtbl.replace oracle key value
    end
    else if op < 75 then begin
      check (D.remove dawg key = present) "crud remove changed";
      Hashtbl.remove oracle key
    end
    else if op < 95 then begin
      check (D.contains dawg key = present) "crud contains";
      if present then
        check ((D.get dawg key).value = Hashtbl.find oracle key) "crud get value"
    end
    else ignore (D.compact dawg);
    check (D.length dawg = Hashtbl.length oracle) "crud size matches oracle"
  done;
  D.close dawg

let c8_substring () =
  let rng = make_rng 0x5CDAL in
  let alphabet = "abcx" in
  let generate max_len =
    let n = rng max_len + 1 in
    String.init n (fun _ -> alphabet.[rng (String.length alphabet)])
  in
  let seen = Hashtbl.create 64 in
  let terms = ref [] in
  while Hashtbl.length seen < 60 do
    let t = generate 6 in
    if not (Hashtbl.mem seen t) then begin Hashtbl.add seen t (); terms := t :: !terms end
  done;
  let terms = !terms in
  let naive pattern =
    List.fold_left
      (fun total term ->
        let count = ref 0 in
        for start = 0 to String.length term - String.length pattern do
          if String.sub term start (String.length pattern) = pattern then incr count
        done;
        total + !count)
      0 terms
  in
  let scdawg = D.scdawg () in
  List.iter (fun term -> ignore (D.put scdawg term None)) terms;
  for _ = 1 to 200 do
    let pattern = generate 3 in
    let expected = naive pattern in
    check (D.substring_frequency scdawg pattern = expected) ("pbt frequency " ^ pattern);
    check (D.contains_substring scdawg pattern = (expected > 0)) ("pbt contains " ^ pattern)
  done;
  D.close scdawg

(* --------------------------------------------------------------------- *)
(* C9 leak discipline                                                    *)
(* --------------------------------------------------------------------- *)

let rss_kib () =
  match (try Some (open_in "/proc/self/status") with _ -> None) with
  | None -> 0
  | Some channel ->
    let result = ref 0 in
    (try
       while true do
         let line = input_line channel in
         if String.length line >= 6 && String.sub line 0 6 = "VmRSS:" then
           (try Scanf.sscanf line "VmRSS: %d" (fun kb -> result := kb) with _ -> ())
       done
     with End_of_file -> ());
    close_in channel;
    !result

let c9 () =
  let cycles = 12000 in
  let batch = [| ("cat", Some 1L); ("cot", Some 2L); ("cut", None) |] in
  for _ = 1 to 2000 do
    let dawg = D.dynamic_dawg () in
    ignore (D.put dawg "cat" (Some 1L));
    D.close dawg
  done;
  Gc.full_major ();
  let before = rss_kib () in
  for _ = 1 to cycles do
    let dawg = D.dynamic_dawg () in
    ignore (D.put_many dawg batch);
    check (D.contains dawg "cot") "leak cycle contains";
    D.close dawg
  done;
  Gc.full_major ();
  let after = rss_kib () in
  if before > 0 && after > before then
    check (after - before < 48 * 1024)
      (Printf.sprintf "RSS grew %d KiB over %d cycles" (after - before) cycles)

(* --------------------------------------------------------------------- *)
(* C10 concurrency (OCaml 5 Domain)                                      *)
(* --------------------------------------------------------------------- *)

let c10_independent () =
  let domains =
    List.init 8 (fun seed ->
      Domain.spawn (fun () ->
        let dawg = D.dynamic_dawg () in
        for i = 0 to 1999 do ignore (D.put dawg (Printf.sprintf "t%d_%d" seed i) (Some (Int64.of_int i))) done;
        let ok = D.length dawg = 2000
                 && (D.get dawg (Printf.sprintf "t%d_1500" seed)).value = Some 1500L in
        D.close dawg;
        ok))
  in
  check (List.for_all Domain.join domains) "independent per-domain dictionaries"

let c10_readers_during_writer () =
  let dawg = D.dynamic_dawg () in
  ignore (D.put_many dawg (Array.init 500 (fun i -> (Printf.sprintf "seed%d" i, Some (Int64.of_int i)))));
  let stop = Atomic.make false in
  let readers =
    List.init 4 (fun _ ->
      Domain.spawn (fun () ->
        let ok = ref true in
        while not (Atomic.get stop) do
          if not (D.contains dawg "seed0") then ok := false;
          ignore (D.get dawg "seed250")
        done;
        !ok))
  in
  for i = 500 to 2999 do ignore (D.put dawg (Printf.sprintf "w%d" i) (Some (Int64.of_int i))) done;
  Atomic.set stop true;
  let all_ok = List.for_all Domain.join readers in
  check all_ok "concurrent readers during writer";
  check ((D.get dawg "w2999").value = Some 2999L) "final write present";
  D.close dawg

(* --------------------------------------------------------------------- *)

let () =
  c1 ();
  c2 ();
  c3 ();
  c4 ();
  c5 ();
  c6 ();
  c7 ();
  c8_crud ();
  c8_substring ();
  c9 ();
  c10_independent ();
  c10_readers_during_writer ();
  if !failures = 0 then print_endline "ocaml conformance: all checks passed"
  else begin
    Printf.eprintf "ocaml conformance: %d check(s) failed\n" !failures;
    exit 1
  end
